internal/tools/aws/client.go

```diff
@@ -51,6 +75,21 @@ type api interface {
 	listTagsForCertificate(*acm.ACM, *acm.ListTagsForCertificateInput) (*acm.ListTagsForCertificateOutput, error)
 }
 
+// NewAWSClient returns a new AWS client.
+func NewAWSClient(sess *session.Session) *Client {
```

Comment:
My bad, I should had stated that I wanted to get a feedback in this part before changing all the rest. Regarding to your first part of the comment, client always needs a session and idealistic we should not be constructing  a new one every time that we make a call. Either client and or session should be initialized once and then passed to the varies parts of the system via context or something similar (idealistically). However, IRC there is a requirement which the application may need to pick up configuration changes from the environment (not sure). I think @gabrieljackson can comment on that.

Reply:
Regarding configuration changes being picked up dynamically, that was regarding using tags to lookup AWS resources and how we wouldn't just want to do that once and then store them all as we could make configuration changes to the account and the already-running provisioner would have stale data afterwards.

This is a bit different in that the changes to session would involve updating the provisioner directly, so I am sorta 0/5 on setting sessions once or as needed for AWS calls.

Reply:
I am also a little confused about the intention of `NewAWSClient` as it relates to the original way to create a client: https://github.com/mattermost/mattermost-cloud/pull/159/files#diff-e359fe546a61f4fa6bd4af55be2c38e5L55

I understand that we will be doing multiple PRs to move to the new mocks+client, but perhaps you could outline how you envision it working.

Reply:
> This is a bit different in that the changes to session would involve updating the provisioner directly, so I am sorta 0/5 on setting sessions once or as needed for AWS calls.

I was referring to a conversation about setting default region in code vs let the session to grab it from the environment. We have two use-cases here that can change the design quite a bit: 

1) setting session once, it's more efficient since a new client and session don't need to be construct every time that we make a call to AWS. The drawback is that the provisioner needs to be restarted always when a configuration changes.

2) creating session on every call, as explained above, it's very inefficient but besides to that, it requires some weird abstractions such are `getClient()` so the code can be tested with a mocked implementation of the api interface. This also makes more difficult for using the interfaces already provided by AWS. If we decided that we really need this use case, the current way how we construct session will need to change since it is currently using [deprecated code](https://github.com/aws/aws-sdk-go/blob/3da93ec4fb4bf582b00a1dc0c5ee87e16d6d446e/aws/session/session.go#L71).

If the system doesn't need to load configuration dynamically without being restarted I think session and client should be constructed once and passed around throughout the whole life-time of the system.

> I am also a little confused about the intention of NewAWSClient as it relates to the original way to create a client: https://github.com/mattermost/mattermost-cloud/pull/159/files#diff-e359fe546a61f4fa6bd4af55be2c38e5L55

I didn't want to remove the old way just yet because of testing and all the call sites that still use the old api interface. The original way to create client requires to manually maintain an implement an interface which the new way doesn't. The new way uses the SDK interfaces to generate mocks from a proper mocking framework. If we decided to go this way I will change all the call sites in this PR and fix the tests that already exist. It's also very likely that I will turn client and session into a singleton.